### PR TITLE
Use Vector<map> instead of vector<deque> for submap_data

### PR DESCRIPTION
**Pull request by jihoonl**
as a preparation work of #409

### CODE REVIEW

**Comment by jihoonl** [comment]
Oh wait. this change might be not necessary. I will reopen it later.

**Comment by jihoonl** [comment]
Reopening the PR as discussed in #409. @wohe  I would like to have a separate PR to remove `num_trimmed_submaps`.  It is used in [ComputeLocalToGlobalTransform](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping_2d/sparse_pose_graph.cc#L566) and [GrowSubmapTransformAsNeeded](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping_2d/sparse_pose_graph.cc#L76) to keep track of the last submap id. So, removing `num_trimmed_submaps` variable would require introducing another variable in optimization_problem. As you prefer small PRs for incremental changes, I would do it in a separate PR.

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 1]
```diff
@@ -144,7 +149,7 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
 
   // Set the starting point.
   // TODO(hrapp): Move ceres data into SubmapData.
-  std::vector<std::vector<std::array<double, 3>>> C_submaps(
+  std::vector<std::map<const int, std::array<double, 3>>> C_submaps(
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: `const` is redundant. Remove?

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 2]
```diff
@@ -153,16 +158,22 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
     const bool frozen = frozen_trajectories.count(trajectory_id);
     // Reserve guarantees that data does not move, so the pointers for Ceres
     // stay valid.
-    C_submaps[trajectory_id].reserve(submap_data_[trajectory_id].size());
-    for (const SubmapData& submap_data : submap_data_[trajectory_id]) {
-      C_submaps[trajectory_id].push_back(FromPose(submap_data.pose));
-      problem.AddParameterBlock(C_submaps[trajectory_id].back().data(), 3);
+    // C_submaps[trajectory_id].reserve(submap_data_[trajectory_id].size());
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: Remove?

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 3]
```diff
@@ -153,16 +158,22 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
...
+    for (auto& it : submap_data_[trajectory_id]) {
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: `it` is not an iterator. `const auto& index_submap_data`?

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 4]
```diff
@@ -153,16 +158,22 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
...
+      const int submap_index = it.first;
+      const SubmapData& submap_data = it.second;
+
+      C_submaps[trajectory_id].insert(
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: `emplace`?

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 5]
```diff
@@ -244,19 +253,16 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
...
 
   // Store the result.
   for (size_t trajectory_id = 0; trajectory_id != submap_data_.size();
        ++trajectory_id) {
-    for (size_t submap_data_index = 0;
-         submap_data_index != submap_data_[trajectory_id].size();
-         ++submap_data_index) {
-      submap_data_[trajectory_id][submap_data_index].pose =
-          ToPose(C_submaps[trajectory_id][submap_data_index]);
+    for (auto& it : submap_data_[trajectory_id]) {
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: Same here: `auto& index_submap_data`?

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 6]
```diff
@@ -244,19 +253,16 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
...
+      submap_data_[trajectory_id][it.first].pose =
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: `index_submap_data.second.pose`?

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 7]
```diff
@@ -244,19 +253,16 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
...
+          ToPose(C_submaps[trajectory_id][it.first]);
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: `.at(index_submap_data.first)`?

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 8]
```diff
@@ -112,18 +112,23 @@ void OptimizationProblem::AddSubmap(const int trajectory_id,
...
 
 void OptimizationProblem::TrimSubmap(const mapping::SubmapId& submap_id) {
   auto& trajectory_data = trajectory_data_.at(submap_id.trajectory_id);
   // We only allow trimming from the start.
   CHECK_EQ(trajectory_data.num_trimmed_submaps, submap_id.submap_index);
   auto& submap_data = submap_data_.at(submap_id.trajectory_id);
-  CHECK(!submap_data.empty());
-  submap_data.pop_front();
+
+  CHECK(submap_data.find(submap_id.submap_index) != submap_data.end());
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: `CHECK` the return value of `erase()`?

File: cartographer/mapping/pose_graph_trimmer.cc [hunk 9]
```diff
@@ -28,12 +28,13 @@ PureLocalizationTrimmer::PureLocalizationTrimmer(const int trajectory_id,
 }
 
 void PureLocalizationTrimmer::Trim(Trimmable* const pose_graph) {
-  const int total_num_submaps = pose_graph->num_submaps(trajectory_id_);
-  while (total_num_submaps > num_submaps_trimmed_ + num_submaps_to_keep_) {
+  int total_num_submaps = pose_graph->num_submaps(trajectory_id_);
+  while (total_num_submaps > num_submaps_to_keep_) {
```
Comment (wohe) [inline cartographer/mapping/pose_graph_trimmer.cc]: `while (pose_graph->num_submaps(trajectory_id_) > num_submaps_to_keep_)`?

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.h [hunk 10]
```diff
@@ -77,7 +77,7 @@ class OptimizationProblem {
              const std::set<int>& frozen_trajectories);
 
   const std::vector<std::deque<NodeData>>& node_data() const;
-  const std::vector<std::deque<SubmapData>>& submap_data() const;
+  const std::vector<std::map<int, SubmapData>>& submap_data() const;
 
   int num_trimmed_nodes(int trajectory_id) const;
   int num_trimmed_submaps(int trajectory_id) const;
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.h]: Can we remove this function and the variable in `TrajectoryData`?

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 11]
```diff
@@ -112,18 +112,23 @@ void OptimizationProblem::AddSubmap(const int trajectory_id,
   CHECK_GE(trajectory_id, 0);
   submap_data_.resize(
       std::max(submap_data_.size(), static_cast<size_t>(trajectory_id) + 1));
-  submap_data_[trajectory_id].push_back(SubmapData{submap_pose});
   trajectory_data_.resize(
       std::max(trajectory_data_.size(), submap_data_.size()));
+
+  auto& trajectory_data = trajectory_data_.at(trajectory_id);
+  submap_data_[trajectory_id].emplace(
+      submap_data_[trajectory_id].size() + trajectory_data.num_trimmed_submaps,
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: Is this the only use of `num_trimmed_submaps` now? Keep track of `trajectory_data.next_submap_index` instead?

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 12]
```diff
@@ -555,20 +550,22 @@ SparsePoseGraph::GetAllSubmapData() {
...
     const std::vector<int>& num_trimmed_submaps,
     const int trajectory_id) const {
   if (trajectory_id >= static_cast<int>(submap_transforms.size()) ||
       submap_transforms.at(trajectory_id).empty()) {
     return transform::Rigid3d::Identity();
   }
-  const mapping::SubmapId last_optimized_submap_id{
-      trajectory_id,
+
+  const int submap_index =
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: Should this now be done using `submap_transforms.at(trajectory_id).rbegin()->first`?

File: cartographer/mapping_2d/sparse_pose_graph.cc [hunk 13]
```diff
@@ -408,7 +402,8 @@ void SparsePoseGraph::AddSubmapFromProto(const int trajectory_id,
   CHECK_EQ(optimized_submap_transforms_.at(trajectory_id).size(),
            submap_id.submap_index);
   optimized_submap_transforms_.at(trajectory_id)
-      .push_back(sparse_pose_graph::SubmapData{initial_pose_2d});
+      .emplace(optimized_submap_transforms_.at(trajectory_id).size(),
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.cc]: RIght now this is the same value, but using `submap_id.submap_index` seems easier to read and more likely to be correct going forward, right?

**Comment by wohe** [comment]
PR size seems to be still okay. If you want to, you could split off the change to `pose_graph_trimmer.cc` and `num_submap()` into a tiny self contained PR.

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 14]
```diff
@@ -153,16 +153,21 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
...
-    C_submaps[trajectory_id].reserve(submap_data_[trajectory_id].size());
-    for (const SubmapData& submap_data : submap_data_[trajectory_id]) {
-      C_submaps[trajectory_id].push_back(FromPose(submap_data.pose));
-      problem.AddParameterBlock(C_submaps[trajectory_id].back().data(), 3);
+    for (const auto& index_submap_data : submap_data_[trajectory_id]) {
+      const int submap_index = index_submap_data.first;
+      const SubmapData& submap_data = index_submap_data.second;
+
+      C_submaps[trajectory_id].emplace(
+          std::pair<const int, std::array<double, 3>>(
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: No need for the `std::pair` anymore with `emplace`?
```
C_submaps[trajectory_id].emplace(submap_index, FromPose(submap_data.pose));
```

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 15]
```diff
@@ -112,19 +112,19 @@ void OptimizationProblem::AddSubmap(const int trajectory_id,
...
 
 void OptimizationProblem::TrimSubmap(const mapping::SubmapId& submap_id) {
-  auto& trajectory_data = trajectory_data_.at(submap_id.trajectory_id);
-  // We only allow trimming from the start.
-  CHECK_EQ(trajectory_data.num_trimmed_submaps, submap_id.submap_index);
   auto& submap_data = submap_data_.at(submap_id.trajectory_id);
-  CHECK(!submap_data.empty());
-  submap_data.pop_front();
-  ++trajectory_data.num_trimmed_submaps;
+  CHECK(submap_data.find(submap_id.submap_index) != submap_data.end());
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: Redundant `CHECK`? Remove.

**Comment by jihoonl** [comment]
ok. done.

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 16]
```diff
@@ -112,19 +112,19 @@ void OptimizationProblem::AddSubmap(const int trajectory_id,
   CHECK_GE(trajectory_id, 0);
   submap_data_.resize(
       std::max(submap_data_.size(), static_cast<size_t>(trajectory_id) + 1));
-  submap_data_[trajectory_id].push_back(SubmapData{submap_pose});
   trajectory_data_.resize(
       std::max(trajectory_data_.size(), submap_data_.size()));
+
+  auto& trajectory_data = trajectory_data_.at(trajectory_id);
+  submap_data_[trajectory_id].emplace(
+      trajectory_data.next_submap_index++,
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: Readability and consistency: Please move the `++trajectory_data.next_submap_index` after the `emplace`.

**Comment by jihoonl** [comment]
Fixed.

File: cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc [hunk 17]
```diff
@@ -153,16 +152,20 @@ void OptimizationProblem::Solve(const std::vector<Constraint>& constraints,
     const bool frozen = frozen_trajectories.count(trajectory_id);
     // Reserve guarantees that data does not move, so the pointers for Ceres
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/optimization_problem.cc]: Remove outdated comment?

**Comment by jihoonl** [comment]
Rebased. or do you want to squash them?

**Comment by jihoonl** [comment]
I reconfirmed that it is fully rebased with master branch.

**Pull request merged**